Clover/app/src/main/java/org/floens/chan/ui/captcha/v2/CaptchaNoJsPresenterV2.java

```diff
@@ -0,0 +1,347 @@
+
+            if (callbacks != null) {
+                callbacks.onCaptchaInfoParseError(error);
+            }
+
+            // return ok here too because we already handled this exception in the callback
+            return RequestCaptchaInfoError.Ok;
+        }
+    }
+
+    private RequestBody createResponseBody(
+            CaptchaInfo prevCaptchaInfo,
+            List<Integer> selectedIds
+    ) throws UnsupportedEncodingException {
+        StringBuilder sb = new StringBuilder();
+
+        sb.append(URLEncoder.encode("c", encoding));
+        sb.append("=");
+        sb.append(URLEncoder.encode(prevCaptchaInfo.getcParameter(), encoding));
+        sb.append("&");
+
+        for (Integer selectedImageId : selectedIds) {
+            sb.append(URLEncoder.encode("response", encoding));
+            sb.append("=");
+            sb.append(URLEncoder.encode(String.valueOf(selectedImageId), encoding));
+            sb.append("&");
+        }
+
+        String resultBody;
+
+        if (selectedIds.size() > 0) {
+            resultBody = sb.deleteCharAt(sb.length() - 1).toString();
+        } else {
+            resultBody = sb.toString();
+        }
+
+        return MultipartBody.create(
+                MediaType.parse(mediaType),
+                resultBody);
+    }
+
+    @Nullable
+    private CaptchaInfo handleGetRecaptchaResponse(Response response) {
+        try {
+            if (response.code() != SUCCESS_STATUS_CODE) {
+                if (callbacks != null) {
+                    callbacks.onCaptchaInfoParseError(
+                            new IOException("Bad status code for captcha request = " + response.code()));
+                }
+
+                return null;
+            }
+
+            ResponseBody body = response.body();
+            if (body == null) {
+                if (callbacks != null) {
+                    callbacks.onCaptchaInfoParseError(
+                            new IOException("Captcha response body is empty (null)"));
+                }
+
+                return null;
+            }
+
+            String bodyString = body.string();
+            if (!bodyString.contains(recaptchaChallengeString)) {
+                throw new IllegalStateException("Response body does not contain \"reCAPTCHA challenge\" string");
+            }
+
+            if (bodyString.contains(verificationTokenString)) {
+                // got the token
+                String verificationToken = parser.parseVerificationToken(bodyString);
+
+                if (callbacks != null) {
+                    callbacks.onVerificationDone(verificationToken);
+                }
+
+                return null;
+            } else {
+                // got the challenge
+                CaptchaInfo captchaInfo = parser.parseHtml(bodyString, siteKey);
+
+                if (callbacks != null) {
+                    callbacks.onCaptchaInfoParsed(captchaInfo);
+                }
+
+                return captchaInfo;
```

Comment:
Probably should return null here to reset prevCaptchaInfo when callbacks are null. Otherwise we are kinda stuck in a state where there is no captcha window (But I guess reset should fix it?).